### PR TITLE
fix(surveys): keep URL prefill working when localStorage is unavailable

### PR DESCRIPTION
**Pull request by darkopia**
## Problem

Hosted (external) surveys opened with a prefilled deep link re-ask the question the link already answered. Reported with a live repro: a thumbs up/down rating question set to "Automatically submit on selection", opened with `?q0=1`, still renders the rating question instead of advancing to the optional follow-up. Same user-visible symptom as #4312, different cause.

The hosted survey page is served with `Content-Security-Policy: sandbox allow-scripts allow-forms allow-popups`, added in [PostHog/posthog#93707](https://github.com/PostHog/posthog/pull/93707) to isolate hosted survey content. Without `allow-same-origin` the document has an opaque origin, so every `localStorage` access throws `SecurityError`. That CSP is deliberate and should stay; this PR adapts the SDK to it rather than relaxing it.

That PR did account for storage being unavailable — it initialises the page with `persistence: 'memory'`, and asserts as much in its tests. The survey in-progress state is what slipped through: it calls `localStorage` directly instead of going through the persistence layer, so the `persistence` setting never applied to it.

Prefill computes the right answer and the right start index, but the in-progress survey state is the only channel carrying them from `renderSurvey` to the question renderer. The write was swallowed, the renderer read nothing back and fell back to question 0. So the respondent is asked again, and the prefilled answer is no longer on the visited path, which drops it from a later manual submit too.

## Changes

The three in-progress state helpers change, all inside the surveys extension:

- a write that throws keeps the state in memory for the page load
- a write that lands drops that copy, so `localStorage` stays authoritative
- a read prefers the in-memory copy when set, which also covers storage modes where writes throw but reads succeed (quota reached, older Safari private browsing)
- clearing a survey clears both, so completion and dismissal leave nothing behind

Wherever storage works this changes nothing, including cross-reload resume and multi-tab reads.

`posthog.reset()` also drops the in-memory copy. Its survey cleanup only scans `localStorage`, so without this the state outlives an identity reset: on a page where storage throws, a logout without a page reload leaves the previous respondent's answers and submission id readable to the next user. Narrow, but reachable in a SPA that resets identity in place. The clear is exposed from the extension and called outside the `localStorage` `try` block, so it still runs when the storage scan itself throws.

It cannot be done in the core directly: `utils/survey-utils` compiles into both the core and the lazily loaded `surveys.js`, so a core-side clear would reach only the core's own unused copy. That was verified in `dist` rather than in jsdom, which collapses the two copies into one and cannot tell them apart — `array.js` calls the method optional-chained on `_surveyManager`, and `surveys.js` defines it over the same map its failed write populates.

**Version skew** (Class 1 of `.agents/skills/replay-incident-risk`): a new core with an older `surveys.js` is safe, because the call is optional-chained and no-ops when the method is absent. An older core with the new `surveys.js` does not get the reset clear, since that core has no such call and the CDN serves the newest extension bundle to pinned SDKs — so the narrow exposure above remains for older cores until they update. Closing it there too would mean keying the map on `distinct_id` inside the extension, which needs no cross-bundle call; worth considering as a follow-up.

Measured against `main`: `surveys.js` 103,956 → 104,139 bytes (+183), `array.js` 312,629 → 312,752 (+123).

### Libraries affected

- [ ] All of them
- [x] posthog-js (web)

## Checklist

- [x] Tests for new code
- [x] Accounted for the impact of any changes across different platforms
- [x] Accounted for backwards compatibility of any changes (no breaking changes!)
- [x] Took care not to unnecessarily increase the bundle size

### If releasing new changes

- [x] Ran `pnpm changeset` to generate a changeset file

## 🤖 Agent context

**Autonomy:** Human-driven (agent-assisted) — DRI: @darkopia

Prefill logic was ruled out first: the reported survey's real config replayed through the unit harness advanced to question 1 correctly, and the deployed bundle already contains the #4312 fix. That pointed at the environment, and the page's response headers showed the `sandbox` CSP. Confirmed in headless Chromium against the reported URL — `SecurityError` from the in-progress state reads, `data-question-index="0"` — then re-run with a locally built patched bundle, which renders the follow-up.

The PR carried a larger `reset()` fix through several commits, was reduced to the prefill fix alone on the maintainer's call, and the `reset()` handling is now restored after review flagged the identity-reset leak. Only that part came back; the jsdom prefill test dropped alongside it stays dropped, since the playwright spec covers the same path against the real bundle. Two review rounds are worth knowing about: the first found that `reset()` bypassed `clearInProgressSurveyState`; the second found that the resulting core-side clear could not work at all, because `utils/survey-utils` compiles into both the core and the lazily loaded `surveys.js`, so it cleared a never-written copy. Both the unit test and the `array.full.js` check were blind to that, since each collapses the two copies into one — the split bundle is the default CDN layout.

Verification: `playwright/mocked/surveys/url-prefill.spec.ts` covers this against the built bundle in a real browser and fails with `data-question-index "0"` without the fix (checked). Full browser unit suite 6850 passing, `tsc --noEmit`, oxlint and oxfmt clean. No manual click-through beyond the automated runs described.

---
*Created with [PostHog](https://posthog.com?ref=pr) from a [Slack thread](https://posthog.slack.com/archives/C0A2AHK6CN8/p1789558069025529)*

### CODE REVIEW

File: packages/browser/src/extensions/surveys/surveys-extension-utils.tsx [hunk 1]
```diff
@@ -777,22 +777,40 @@ const getInProgressSurveyStateKey = (survey: Pick<Survey, 'id' | 'current_iterat
...
 ): void => {
+    const key = getInProgressSurveyStateKey(survey)
     try {
-        localStorage.setItem(getInProgressSurveyStateKey(survey), JSON.stringify(state))
+        localStorage.setItem(key, JSON.stringify(state))
+        // The write landed, so drop any copy left by an earlier failed one.
+        delete inMemoryInProgressSurveyState[key]
     } catch (e) {
         logger.error('Error setting in-progress survey state in localStorage', e)
+        inMemoryInProgressSurveyState[key] = state
```
[the remark this answers is not in the document]
Comment (dustinbyrne) [reply packages/browser/src/extensions/surveys/surveys-extension-utils.tsx:803]: my agent also flagged this. it looks like we were clearing the in-memory cache on `reset`, but it was removed in https://github.com/PostHog/posthog-js/commit/313a6ded34244475d4e93f5d0d8780ce5d5aa390

seems worth retaining that? i wouldn't revert that commit outright because it looks like the test case removal in packages/browser/src/__tests__/extensions/surveys.test.ts was valid and duplicate

super edge case-y because this is only memory persistence but i suppose it could affect a SPA

**Review by dustinbyrne — APPROVED** [review]
